src/core_write.cpp

```diff
@@ -109,33 +115,46 @@ std::string ScriptToAsmStr(const CScript& script, const bool fAttemptSighashDeco
             str += "[error]";
             return str;
         }
+        // Covers OP_FALSE
+        if (opcode == OP_0) {
+            str += "0";
+            continue;
+        }
+        if (opcode == OP_1NEGATE) {
+            str += "-1";
+            continue;
+        }
+        // OP_n -> n (so 1, ..., 10, ..., 16), covers OP_TRUE
+        if ((opcode >= OP_1 && opcode <= OP_16)) {
+            // These are always minimal, use the corresponding value
+            str += strprintf("%i", opcode - OP_1NEGATE - 1);
+            continue;
+        }
         if (0 <= opcode && opcode <= OP_PUSHDATA4) {
-            if (vch.size() <= static_cast<std::vector<unsigned char>::size_type>(4)) {
-                str += strprintf("%d", CScriptNum(vch, false).getint());
-            } else {
-                // the IsUnspendable check makes sure not to try to decode OP_RETURN data that may match the format of a signature
-                if (fAttemptSighashDecode && !script.IsUnspendable()) {
-                    std::string strSigHashDecode;
-                    // goal: only attempt to decode a defined sighash type from data that looks like a signature within a scriptSig.
-                    // this won't decode correctly formatted public keys in Pubkey or Multisig scripts due to
-                    // the restrictions on the pubkey formats (see IsCompressedOrUncompressedPubKey) being incongruous with the
-                    // checks in CheckSignatureEncoding.
-                    if (CheckSignatureEncoding(vch, SCRIPT_VERIFY_STRICTENC, nullptr)) {
-                        const unsigned char chSigHashType = vch.back();
-                        const auto it = mapSigHashTypes.find(chSigHashType);
-                        if (it != mapSigHashTypes.end()) {
-                            strSigHashDecode = "[" + it->second + "]";
-                            vch.pop_back(); // remove the sighash type byte. it will be replaced by the decode.
-                        }
-                    }
-                    str += HexStr(vch) + strSigHashDecode;
+            if (CheckMinimalPush(vch, opcode)) {
+                if (vch.size() <= 5 ) {
+                    // Return decimal for minimially-encoded values <= 5 bytes (OP_CLTV / OP_CSV accept 5-byte numbers)
+                    CScriptNum n(vch, false, 5);
```

Comment:
Agree that it should be `true` as a belt-and-suspenders, but we are inside `if (CheckMinimalPush(vch, opcode))` already here, so shouldn't happen?

Reply:
Byte code `024000` is a minimal push of the two-byte vector `4000` (so passes `CheckMinimalPush`) but it's not a minimal `CScriptNum` since it just evaluates to 64, and should have been `0140` (so fails `fRequireMinimal` due to the trailing 0).

Reply:
Thanks for the explanations.

I've refactored out handling pushed data, as I wasn't enjoying all the nested if statements, and notice that we are currently handling "non-minimal pushes" the same as minimal pushes with non-minimally-encoded values, which is to use `OPCODE<hex>`. This seems ok, and it will not be a lossy decode, but do we want to differentiate between the two pathways to this decode format?

See first and last case in snippet to illustrate what I mean above:

```cpp
/*
 * Format pushed bytes into appropriate ASM format
 */
std::string FormatPushDataAsm(const std::vector<unsigned char>& vch, opcodetype opcode)
{
    // Use OPCODE<hex> for non-minimal pushes
    // TODO: There are no tests for this currently
    if (!CheckMinimalPush(vch, opcode)) {
        return GetOpNameAsm(opcode) + BracketStr(HexStr(vch));
    }

    // Use <hex> for minimal pushes > 5 bytes
    if (!(vch.size() <= 5) ) {
        return BracketStr(HexStr(vch));
    }

    // Use decimal for minimally-encoded, minimal pushes <= 5 bytes
    // Note: OP_CLTV / OP_CSV accept 5-byte numbers
    try {
        CScriptNum n{vch, true, 5};
        return strprintf("%lld", n.GetInt64());
    }

    // Use OPCODE<hex> for non-minimally-encoded minimal pushes
    // TODO: There are no tests for this currently
    catch (scriptnum_error& e) {
        return GetOpNameAsm(opcode) + BracketStr(HexStr(vch));
    }
}
```

Reply:
Sure, but what I mean is that in my current impl there are _two_ ways we can end up with `OPCODE<...>`:

1) Wasn't a minimal push
2) Was a minimal push <5 bytes, but wasn't a minimally-encoded decimal value

Reply:
Yeah. This just means changing the final `catch` to `return BracketStr(HexStr(vch));`, no?

Reply:
Also,

```c++
if (!(vch.size() <= 5) ) {
```

Can I introduce you to `>` ? :smile: